Add form-specific SubmitHistory report link to form designer

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

This PR adds a "View Submissions" button to the form designer that will take user to the form-specific submission history. It dynamically generates a filtered SubmitHistory report URL with form-specific parameters and date range filtering.
This is aimed to provide new user a easy way to find the data they submitted.

"View Submissions" button on a app settings page -- Submission History will filter by app
<img width="1251" alt="image" src="https://github.com/user-attachments/assets/cbfcd257-3ac5-4f90-ada6-379fd75b37be" />

"View Submissions" button on a module settings page -- Submission History will filter by app and module
<img width="1256" alt="image" src="https://github.com/user-attachments/assets/0429795d-3b97-441d-9635-cd7501b4302b" />

"View Submissions" button on a form settings page -- Submission History will filter by app, module and form
<img width="1251" alt="image" src="https://github.com/user-attachments/assets/87461fbe-c12e-41f8-953d-94e3286eadf0" />

"View Submissions" button on a form editing page -- Submission History will filter by app, module and form
<img width="1255" alt="image" src="https://github.com/user-attachments/assets/6861a125-75d8-4e8c-8c99-7b0aa9b29d42" />


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-17723

The technical change is mainly generating the link in backend in the function `get_form_submit_history_url_for_last_30_days`, then pass the url as context to all the views that has the "View Submissions" button partial template.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally, button show up correctly, the link generated correctly, the filter set up correctly.
This change only adds a link to view existing data, so I feel it is safe.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->



### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Add View Submissions Button in case list's form editing page

jingcheng16 committed:
Show View Submissions button in Survey Form

jingcheng16 committed:
Create another View Submission button partial for white background and rename the previous partial to ..._on_dark_bg

jingcheng16 committed:
Make the function public and more flexible to accept parameters being None

jingcheng16 committed:
Add View Submission button to app setting, module setting, form setting page

jingcheng16 committed:
Refactor: create a new permission constant to avoid hardcoding it

jingcheng16 committed:
Refactor: Make make_url a util function

jingcheng16 committed:
Leave space for new buttons floating on the right

jingcheng16 committed:
Fix failed test since I add new context

gherceg commented on corehq/apps/reports/util.py:
```diff
@@ -1070,3 +1071,10 @@ def get_commcare_version_and_date_from_last_usage(last_submission=None, last_dev
+
+
+def make_url(base_url, params):
+    return '{base_url}?{params}'.format(
+        base_url=base_url,
+        params=urlencode(params, True),
+    )
```
nit: I like f-strings 
```suggestion
    return f'{base_url}?{urlencode(params, True)}'
```

jingcheng16 replied:
End up using `reverse` in f2c6a12154d8316d3cc13cccf9cbd88275289775

nospame commented on corehq/apps/app_manager/views/formdesigner.py:
```diff
@@ -98,6 +101,33 @@ def form_source_legacy(request, domain, app_id, module_id=None, form_id=None):
+    return _make_url(base_url, params)
+
+
+def _make_url(base_url, params):
```
I believe you can use `corehq.util.view_utils` `reverse` for this https://github.com/dimagi/commcare-hq/blob/e9cef132ecf35a20ccc3b5dcb955b31a5fbe46c5/corehq/util/view_utils.py#L119

gherceg reviewed (approved):
A couple of comments but nothing blocking.

gherceg committed:
Merge branch 'master' into jc/add-view-submissions-button

jingcheng16 committed:
Revert "Refactor: Make make_url a util function"

This reverts commit 897409b81f95df80baade0020c00d8b6ae924578.

jingcheng16 committed:
Use reverse instead of make_url

jingcheng16 committed:
Refactor: Rename VIEW_SUBMISSION_HISTORY_PERMISSION to SUBMISSION_HISTORY_PERMISSION

jingcheng16 commented on corehq/apps/users/permissions.py:
```diff
@@ -5,6 +5,8 @@
 from corehq.apps.accounting.utils import domain_has_privilege
 
+VIEW_SUBMISSION_HISTORY_PERMISSION = 'corehq.apps.reports.standard.inspect.SubmitHistory'
+
```
Make sense!  4c4377fada8941de37e844efe0ffddd5a30723db

jingcheng16 committed:
Use one template and dynamically set the button class

jingcheng16 commented on corehq/apps/hqwebapp/static/app_manager/less/content.less:
```diff
@@ -157,10 +157,10 @@
 .appmanager-edit-title {
-  margin-right: 300px;
+  margin-right: 400px;
 }
 .appmanager-edit-app-title {
-  margin-right: 248px;
+  margin-right: 400px;
```
😂 Just leaving roughly "enough" space

jingcheng16 committed:
Bad form can miss xmlns

This is to fix the test failure in https://github.com/dimagi/commcare-hq/actions/runs/15865661315/job/44731936498?pr=36606

jingcheng16 committed:
Add context key

This is to fix error in https://github.com/dimagi/commcare-hq/actions/runs/15865661315/job/44731936496?pr=36606

nospame commented on corehq/apps/app_manager/views/formdesigner.py:
```diff
@@ -102,13 +101,14 @@ def form_source_legacy(request, domain, app_id, module_id=None, form_id=None):
 
 
 def get_form_submit_history_url_for_last_30_days(request, domain, app, module, form):
+    from corehq.util.view_utils import reverse
```
Did it cause circular import issues to have this at the top of the file, or is it just the name conflict with `django.urls.reverse`? If it's the latter, I would prefer to see it renamed during import, something like this, instead of the local import https://github.com/dimagi/commcare-hq/blob/c3a157e43b63a78fa4e323d3e27b764d73ec1723/corehq/apps/data_interfaces/views.py#L95

nospame commented on corehq/apps/app_manager/templates/app_manager/app_view.html:
```diff
@@ -8,7 +8,7 @@
   <div class="appmanager-page-actions">
     {% include 'app_manager/partials/app_summary_button.html' %}
-    {% include 'app_manager/partials/view_submissions_button_on_white_bg.html' %}
+    {% include 'app_manager/partials/view_submissions_button.html' with btn_style='btn-default' %}
```
Just confirming you've tested this works without explicitly passing `form_submit_history_url` here?

jingcheng16 replied:
Yes! tested locally

nospame commented on corehq/apps/app_manager/views/formdesigner.py:
```diff
@@ -113,7 +113,7 @@ def get_form_submit_history_url_for_last_30_days(request, domain, app, module, f
         'form_app_id': app.id if app else '',
         'form_module': module.id if module else '',
-        'form_xmlns': form.xmlns if form else '',
+        'form_xmlns': form.xmlns if form and form.xmlns else '',
```
Could also do `getattr(form, 'xmlns', '') if form else ''` but don't feel strongly about it.

jingcheng16 committed:
Rename the function during import to avoid conflicts
